tsweb/jsonhandler.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"encoding/json"
+	"net/http"
+	"reflect"
+)
+
+// NewJSONHandler wraps an HTTP handler function with a version that automatically
+// unmarshals and marshals requests and responses respectively into fn's arguments
+// and results.
+// fn must have a signature with 2 or 3 input arguments. The first two must be http.ResponseWriter and
+// *http.Request. The optional third argument can be of any type.
+// fn's result parameter can be either (error) or (T, error), where T is the JSON-marshalled result type.
+func NewJSONHandler(fn interface{}) http.Handler {
+	v := reflect.ValueOf(fn)
+	t := v.Type()
+	// reflect.Call games
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		wv := reflect.ValueOf(w)
+		rv := reflect.ValueOf(r)
+		var vs []reflect.Value
+
+		switch t.NumIn() {
+		case 2:
+			vs = v.Call([]reflect.Value{wv, rv})
+		case 3:
+			dv := reflect.New(t.In(2))
+			err := json.NewDecoder(r.Body).Decode(dv.Interface())
+			if err != nil {
+				w.WriteHeader(http.StatusBadRequest)
```

Comment:
Write a JSON response here that includes the decode error, to help poor souls (me) when they make a mistake writing a `curl` command.